### DISCUSSION TITLE
@customElement and tree shaking

### DISCUSSION THREAD

**Discussion by ryantrem** [lit/lit#4772] [Q&A]
From what I can tell, a common practice with custom elements is to call `customElements.define` at module scope, making it a side effect of loading the module. Lit pretty much follows suit though automates this a bit with the `@customElement` decorator. However, since this is a side effect, it will often be tree shaken away by bundlers. I would think that to avoid this you would need to either:
1. Mark the specific module (file) that defines the custom element has having side effects in your package.json and rely on the bundler to respect this.
2. Export a function that must be called explicitly to register the element (e.g. a function that calls `customElements.define`).

However, I don't see popular projects that use Lit do either of these things. Any insights here? What are common "correct" patterns for this?

**Comment by yhaiovyi-tripactions** [comment]
I have the exact same problem and don't see any good solution at the moment.

**Comment by polerin** [comment]
I've seen something work without it, but I wasn't able to distinguish exactly why.  That said, my current practice is to have a function that basically does a `MyCustomElement.prototype` for all of the Lit elements, then call that from or close to the root of the application.

**Comment by dufusproxy** [comment]
I am new to Lit.

But my personal solution was to create decorator to use instead of the Lit `@customElement` directly.
It allows you to defer the call thus eliminating the side effect.

```ts
import { LitElement, html } from 'lit'
import { registerCustomElement, type RegisterableConstructor } from '.registerCustomElement.js'

@registerCustomElement('my-component')
class MyComponent extends LitElement {
  render() { html`<slot></slot>` }
}

// this will make the type for `.register` work
export default MyComponent as typeof MyComponent & RegisterableConstructor
```

```ts
import MyComponent from './my-component.ts'

MyComponent.register()
```

https://gist.github.com/dufusproxy/a3e749c9d17cc60fdb0a4c594e40cace

**Reply by dufusproxy** [reply]
@ryantrem sorry, I was wrong with this... it still useful for choosing a different element name I guess, but my assumption on tree shaking was wrong.

My assumption was that the side effect would cause all components or an export file to appear in build if you only imported one when tree shaking is turned on. I asked AI and it confirmed my assumption.

When I tested, it that wasn't the case the tree shaking worked fine - thanks AI lol.

When I re-read your first post I realised you said:
> since this is a side effect, it will often be tree shaken away by bundlers

I don't understand what that means, wouldn't it only be tree shaken away if not used?
So I doubt my idea would help whatever you are trying to solve.

**Reply by ryantrem** [reply]
Normally you register a custom element with custom elements.define. I'm assuming Lit's decorator is resulting in module level code that calls customElements.define. Any module level code that could affect State outside the module is considered a side effect. Therefore, if sideEffects is set to false in the package.json, the bundler will assume that it can tree shake away that module level code and therefore the entire custom element if it is not explicitly referenced in app code (eg if the custom element is only used in html). That said, I've seen some packages that don't set sideEffects to true and somehow the custom Lit-based element still works. I expect omitting the sideEffects flag entirely is the same as explicitly seeing it to true for bundler backward compatibility.

**Reply by dufusproxy** [reply]
this is a build I did when i was testing that idea...
it seems to have everything required with sideEffects set false

> if it is not explicitly referenced in app code (eg if the custom element is only used in html)

But custom elements only work if you explicitly reference them right? they need to be there and you need to call `customElements.define` or `customElement` in the case of lit (same thing but a little bit extra stuff) if you just use `<my-component>` in html without defining it wont work (well it wont be the component just a foreign unknown element)

Not being argumentative just curious :)

<img width="1372" height="1115" alt="image" src="https://github.com/user-attachments/assets/4fdb5f0e-8153-4eab-9de7-8e07c229d31a" />

**Reply by dufusproxy** [reply]
This is when i remove `"sideEffects": false` from package.json

<img width="1367" height="1093" alt="image" src="https://github.com/user-attachments/assets/dad15e17-60ca-45f0-9bb8-c61e37e9557e" />

The only difference is the traditional-library.js has unused components

**Reply by dufusproxy** [reply]
Oh wait do you mean if you just do `import 'my-component.ts'` given theres no reason to import `MyComponent` unless you want to use it to extend another component etc?
